homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,324 @@
+                                          precision=precision,
+                                          entity=entity_id))
+
+    async_add_devices([SensorFilter(name, entity_id, filters)])
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    def __init__(self, name, entity_id, filters):
+        """Initialize the sensor."""
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+        self._icon = ICON
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+            self._icon = new_state.attributes.get(ATTR_ICON, ICON)
+
+            self._state = new_state.state
+
+            if self._state == STATE_UNKNOWN:
+                return
+
+            for filt in self._filters:
+                try:
+                    filtered_state = filt.filter_state(self._state)
+                    _LOGGER.debug("%s(%s, %s) -> %s", filt.name,
+                                  self._entity,
+                                  self._state,
+                                  "skip" if filt.skip else filtered_state)
+                    if filt.skip:
+                        return
+                except ValueError:
+                    _LOGGER.warning("Could not convert state: %s to number",
+                                    self._state)
+                finally:
+                    self._state = filtered_state
+                    filt.states.append(filtered_state)
+
+            self.async_schedule_update_ha_state(True)
+
+        async_track_state_change(
+            self.hass, self._entity, filter_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement of the device."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        state_attr = {
+            ATTR_ENTITY_ID: self._entity
+        }
+        for filt in self._filters:
+            for filt_stat_key, filt_stat_value in filt.stats.items():
+                filt_stat = slugify("{} {}".format(filt.name, filt_stat_key))
```

Comment:
Is it really necessary to add all this data to the state machine? That's a lot of info.

Also, why call slugify and not just do `"{}_{}".format(filt.name, filt_stat_key)`

Reply:
decided with slugify because I'm using name, and was not sure what can come in there...

Should I move to string format only ?

Reply:
you only need slugify if the incoming data might not be appropriate for being stored in the state attributes. Since you control the filters, you can just make the values not require any slugification.